### DISCUSSION TITLE
Plans to support for SSG/ISR?

### DISCUSSION THREAD

**Discussion by nghiepdev** [remix-run/remix#2796] [Proposals]
Hey team,

Are there plans to support SSG/ISR?

**Comment by sergiodxa** [comment]
There was a discussion from before v1 here https://github.com/remix-run/remix/issues/179

I don’t think there’s an immediate plan from the team to add SSG/ISR support, you could get the same results by using Cache Control headers in your routes, Next’s ISR is a Cache Control with the stale-while-revalidate directive with a custom and vendor lock-in API, while the Cache Control header is standard.

**Comment by jdgamble555** [comment]
That discussion link only briefly mentions ISR.  This should be added to the Framework, as I believe this is a huge reason certain people would prefer NextJS over Remix.

The [Remix Team](https://remix.run/blog/remix-vs-next) claims that ISR is vendor-lock-in, but this is simply not true.  Netlify, Firebase, Heroku, Cloud Run, Digital Ocean, etc all support ISR on NextJS.  Not only that, [NuxtJS](https://www.vuemastery.com/blog/new-nuxt-3-feature-incremental-static-generation/) has added the feature, and Rich Harris has said SvelteKit will have the feature after 1.0.

This is huge as it will not only make revalidation easy (would need an adapter for different environments), but it would save database reads, and be faster.  

Please add this ability (or something similar to easily use and revalidate `cache-control`.

J

**Reply by sergiodxa** [reply]
Most CDNs have an API endpoint to purge cached responses, that endpoint depends on your CDN and Remix has no opinion on that so you should check yourself what CDN you're using and how to do that. Then you can use Cache-Control to cache forever or enable SWR.

Next.js ISR support outside Vercel doesn't work the same as in Vercel, because in Vercel your static HTMLs are cached in their CDN, when using for example Heroku or Digital Ocean your static HTML files are cached on disk AFAIK. I think only Netlify has added support and it's something specific for Next, ISR is not a feature all frameworks can do regardless of where you deploy.

**Reply by jdgamble555** [reply]
So I would respond to that by asking if Windows makes you "check yourself" on how to install drivers based on your individual hardware? 
These days it just installs your drivers for you, and you don't have to think about it.  One option would definitely be to put this on the users. 

However, a better option would be to create an adapter pattern so that each Hosting Environment Company and environment can create their own adapter that works seamlessly without the user having to think about it.  This is what Vercel did with NextJS and will do with SvelteKit.  This is how NuxtJS works.

How the CDN stores static files is of little importance compared to the ability to cache them in the first place, and cache them easily because the frameworks handles this well.  By putting the standard into the framework, you will open up the door for the framework to prosper among more users.

J

**Reply by airtonix** [reply]
It's a nice idea, and one worth following, but this:

> So I would respond to that by asking if Windows makes you "check yourself" on how to install drivers based on your individual hardware?
> These days it just installs your drivers for you, and you don't have to think about it. One option would definitely be to put this on the users.

Made me cringe.

A lot of developers use Linux, where you are indeed required to think about these kinds of things.

I think if you're trying to create pressure by comparing common scenarios, then your following statements about how nuxt/next etc work would be enough.

**Comment by ascorbic** [comment]
I've written a guide on how to do ISR on Remix with normal cache headers. It's not _exactly_ the same as ISR on Next.js, but it gives you all the good bits: on-demand revalidation, TTL, stale-while-revalidate. https://developers.netlify.com/guides/how-to-do-isr-and-advanced-caching-with-remix/

**Comment by candidia** [comment]
This is a great API idea https://x.com/kettanaito/status/1748670984381419736

**Comment by candidia** [comment]
Also it's coming https://x.com/brophdawg11/status/1747402216925704684

**Comment by raRaRa** [comment]
Any update on this since single fetch has landed? 🙏

**Reply by sergiodxa** [reply]
The team is working on that https://twitter.com/ryanflorence/status/1788685315462631426

**Reply by airtonix** [reply]
@sergiodxa For those that don't have/use twitter, is there a github issue that is tracking this progress?

**Reply by sergiodxa** [reply]
https://github.com/remix-run/react-router/pull/11539 it's merged on RR, I think it will be added on RRv7 directly and not on Remix

**Reply by sergiodxa** [reply]
You can also see what the team is working on here https://github.com/orgs/remix-run/projects/5

**Comment by quantuminformation** [comment]
great news guys, saves me needing to learn astro
